servers/zms/src/main/java/com/yahoo/athenz/zms/QuotaChecker.java

```diff
@@ -178,49 +194,93 @@ void checkGroupQuota(ObjectStoreConnection con, String domainName, Group group,
         }
     }
 
-    void checkRoleMembershipQuota(ObjectStoreConnection con, String domainName,
-            String roleName, String caller) {
-        
-        // if quota check is disabled we have nothing to do
-        
-        if (!quotaCheckEnabled) {
+    void checkRoleMembershipQuota(ObjectStoreConnection con, final String domainName,
+            final String roleName, final String memberName, Integer maxMembers, final String caller) {
+
+        // if quota check is disabled or the max member limit is not set
+        // on the role then we have nothing to do
+
+        if (!quotaCheckEnabled && (maxMembers == null || maxMembers == 0)) {
             return;
         }
-        
-        // first retrieve the domain quota
+
+        // we're going to check if the current member is already either
+        // a standard or pending member in the role. If that's the case
+        // then we don't need to enforce the quota since we're modifying
+        // an existing entry
+
+        Membership membership = con.getRoleMember(domainName, roleName, memberName, 0, false);
+        if (membership.getIsMember()) {
+            return;
+        }
+
+        // so at this point we know that we'll be adding a new member to the
+        // role. so first let's count the number of role members
+
+        int roleMemberCount = con.countRoleMembers(domainName, roleName);
```

Comment:
Is it meant to count the expired members also as valid role/group member in terms of count?

Reply:
yes, the count includes everyone regardless if they're expired or not